### PR TITLE
faster PileupElement.isBefore and isAfter queries with no linked list allocation.

### PR DESCRIPTION
**Pull request by akiezun**
those came up clearly on HaplotypeCaller profiles and allocated many megabytes of objects for now reason

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 1]
```diff
@@ -351,11 +346,27 @@ public int getOffsetInCurrentCigar() {
             }
         }
 
-        // optimization: elements is null because nothing got added, just return the empty list
         return elements == null ? new LinkedList<>() : elements;
     }
 
     /**
+     * Helper function to get cigar elements between this and either the prev or next genomic position
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: hmm doc needs fixing

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 2]
```diff
@@ -351,11 +346,27 @@ public int getOffsetInCurrentCigar() {
...
+     *
+     * @param direction PREVIOUS if we want before, NEXT if we want after
+     * @return the neighboring position in given direction or null if there is none.
+     */
+    private final CigarElement getNeighborElement(final Direction direction) {
+        final int increment = direction == Direction.NEXT ? 1 : -1;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Increment should be baked into the enum.

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 3]
```diff
@@ -351,11 +346,27 @@ public int getOffsetInCurrentCigar() {
...
+
+        final List<CigarElement> cigarElements = read.getCigarElements();
+        final int i = currentCigarOffset + increment;
+        if (i >= 0 && i < cigarElements.size()){
+            return cigarElements.get(i);
+        } else {
+            return null;
+        }
+    }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Unit test for this method (make package accessible)

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Check `ON_GENOME_OPERATORS`?

**Comment by lbergelson** [comment]
@akiezun Is there a reason these are `LinkedLists`?  It seems like it should be List instead.  I tried to find usages that depend on the linkedlistness of it, but the only thing I see is that one private method relies on it being a `Deque`, but that could be trivially rewritten.   It seems like you'd avoid even more allocation by switching it to use List and then using the non-allocating `Collections.emptyList()`

`getBetween` uses the linkedness of it to use `addFirst`.  This could be rewritten to either build the list in the reverse order and reverse it, or maybe provide a guava `Lists.reverse` view.

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 4]
```diff
@@ -415,30 +426,58 @@ private boolean hasOperator(final CigarElement maybeCigarElement, final CigarOpe
     /**
      * Does an insertion occur immediately before the current position on the genome?
      *
-     * @return true if yes, false if no
+     * @return true if an insertion occur immediately before the current position on the genome, false otherwise
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: should be `occurs` in all of these

**Comment by lbergelson** [comment]
There's a very suboptimal operation in `getBetween`  It calls `read.getCigar().getCigarElement(i);` once for every element. `getCigar()` is an expensive operation that makes a copy every time.

**Comment by droazen** [comment]
Address current comments and then I/we will re-review

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: done

**Comment by akiezun** [comment]
back to you - i removed the linked list type now. it used to be needed but no more. added some tests for new methods.

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 5]
```diff
@@ -317,45 +317,69 @@ public int getOffsetInCurrentCigar() {
...
     }
 
     /**
+     * Helper function to get cigar operator right next to this position but only for not-on-genome operators.
+     *
+     * @param direction PREVIOUS if we want before, NEXT if we want after
+     * @return the neighboring position in given direction or null if there is none
+     */
+    @VisibleForTesting
+    CigarOperator getNeighborNotOnGenomeOperator(final Direction direction) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: `getNextNotOnGenomeOperator(Direction)` reads a little less awkwardly

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 6]
```diff
@@ -317,45 +317,69 @@ public int getOffsetInCurrentCigar() {
...
+            Collections.reverse(elements);
+        }
+        return elements;
     }
 
     /**
+     * Helper function to get cigar operator right next to this position but only for not-on-genome operators.
+     *
+     * @param direction PREVIOUS if we want before, NEXT if we want after
+     * @return the neighboring position in given direction or null if there is none
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: `the next not-on-genome Cigar operator in the given direction, or null if there is none`

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 7]
```diff
@@ -387,7 +411,7 @@ public CigarElement getNextOnGenomeCigarElement() {
      * @return a CigarElement, or null if no such element exists
      */
     private CigarElement getNeighboringOnGenomeCigarElement(final Direction direction) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Make sure the name of this method is consistent with whatever name you end up with for the new method above.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Wait a sec...this method `getNeighboringOnGenomeCigarElement()` behaves very differently from `getNeighborNotOnGenomeOperator()`, since it iterates until it finds an "on-genome" operator, whereas your new method does not iterate. This is a somewhat confusing state of affairs -- can you clean this up to increase clarity?

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 8]
```diff
@@ -317,45 +317,69 @@ public int getOffsetInCurrentCigar() {
...
     /**
      * Helper function to get cigar elements between this and either the prev or next genomic position
      *
      * @param direction PREVIOUS if we want before, NEXT if we want after
      * @return a non-null list of cigar elements between this and the neighboring position in direction
      */
-    private LinkedList<CigarElement> getBetween(final Direction direction) {
-        final int increment = direction == Direction.NEXT ? 1 : -1;
-        LinkedList<CigarElement> elements = null;
+    private List<CigarElement> getBetween(final Direction direction) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Document the behavior of this confusing method -- especially the fact that it stops accumulating elements at the first "on-genome" operator it encounters

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 9]
```diff
@@ -415,35 +439,59 @@ private boolean hasOperator(final CigarElement maybeCigarElement, final CigarOpe
...
+    }
+
+    /**
+    * @return true if a given operator event occurs immediately after the current position on the genome, false otherwise.
      */
-    public boolean isBeforeSoftClip() { return isBefore(getBetweenNextPosition(), CigarOperator.S); }
+    @VisibleForTesting
+    boolean isImmediatelyBefore(final CigarOperator op) {
+        Utils.nonNull(op);
+        return atEndOfCurrentCigar() && getNeighborNotOnGenomeOperator(Direction.NEXT) == op;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: `isAfter()` didn't have this `atEndOfCurrentCigar()` check -- are you sure that your new code is equivalent to the previous version?

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java [hunk 10]
```diff
@@ -55,6 +56,16 @@ public ReadPileup(final Locatable loc, final List<GATKRead> reads, final List<In
     }
 
     /**
+     * Returns the element corresponding to the given read of null there is no such element.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: of -> or

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java [hunk 11]
```diff
@@ -55,6 +56,16 @@ public ReadPileup(final Locatable loc, final List<GATKRead> reads, final List<In
...
+     * @param read or null if elements with no reads are to be retrieved from this pileup.
+     * @return
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: Complete your javadoc

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java [hunk 12]
```diff
@@ -55,6 +56,16 @@ public ReadPileup(final Locatable loc, final List<GATKRead> reads, final List<In
     }
 
     /**
+     * Returns the element corresponding to the given read of null there is no such element.
+     * @param read or null if elements with no reads are to be retrieved from this pileup.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: Can there be an element with no read? What if there are multiple such elements? (document)

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 13]
```diff
@@ -155,6 +157,8 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
         Assert.assertEquals(pe2.getBetweenPrevPosition().size(), ops.size());
         Assert.assertEquals(pe2.getBetweenNextPosition().size(), 0);
         assertEqualsOperators(pe2.getBetweenPrevPosition(), ops);
+        Assert.assertEquals(pe2.getNeighborNotOnGenomeOperator(PileupElement.Direction.PREV), ops.get(ops.size()-1));
+        Assert.assertEquals(pe2.getNeighborNotOnGenomeOperator(PileupElement.Direction.NEXT), null);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: These tests for the new `getNeighborNotOnGenomeOperator()` method do not really seem adequate -- can we get a high-quality standalone unit test that properly exercises the method?

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 14]
```diff
@@ -147,6 +147,8 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
         Assert.assertEquals(pe.getBetweenPrevPosition().size(), 0);
         assertEqualsOperators(pe.getBetweenNextPosition(), ops);
         Assert.assertEquals(pe.getPreviousOnGenomeCigarElement(), null);
+        Assert.assertEquals(pe.getNeighborNotOnGenomeOperator(PileupElement.Direction.PREV), null);
+        Assert.assertEquals(pe.getNeighborNotOnGenomeOperator(PileupElement.Direction.NEXT), ops.get(0));
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: The neighboring "not on genome" operator is always the first element in ops? What if `ops.get(0)` is an "on genome" operator?

**Comment by droazen** [comment]
Can you please test this change with the `HaplotypeCaller` in protected and make sure nothing changes? In particular, can you run `HaplotypeCallerIntegrationTest` and `HaplotypeCallerEngineUnitTest` and make sure they pass? This PR makes me a little nervous given the centrality of the classes touched, even though the optimization itself is simple enough...

**Comment by droazen** [comment]
Back to @akiezun

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 13 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Also, where are the tests for `isImmediatelyBefore/After()`? They are `@VisibleForTesting` but not tested!

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 15]
```diff
@@ -415,35 +439,59 @@ private boolean hasOperator(final CigarElement maybeCigarElement, final CigarOpe
...
      *
-     * @return true if yes, false if no
+     * @return true if a soft-clipping event occurs immediately after the current position on the genome, false otherwise.
+     */
+    public boolean isBeforeSoftClip() {
+        return isImmediatelyBefore(CigarOperator.S);
+    }
+
+    /**
+     * @return true if a given operator event occurs immediately before the current position on the genome, false otherwise.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: These `isImmediatelyBefore/After` methods will not work at all if the `op` is an on-genome operator, but this is totally undocumented and unexpected.

**Comment by akiezun** [comment]
hmm, this is turning into a major piece of work. Maybe i'll redo it from scratch and keep minimal.

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: renamed the other method to `getAdjacentNotOnGenomeOperator` to indicate that it's the immediate neighbor

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: added this text verbatim

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java (hunk 10 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java (hunk 11 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java (hunk 12 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java]: done. it returns the first one it finds or null if it finds none

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 13 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: it;s a bit painful to test in separation - that's why it was private before

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: added tests for isImmediatelyBefore/After

Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: extracted the tests to own method - let me know if this is better for you

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 14 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: that can't be -  by the way the test is contructed the `ops` list only contains `C` `P` and `S`

**Comment by akiezun** [comment]
FWIW HaplotypeCaller tests pass with these changes

**Comment by akiezun** [comment]
back for rereview @droazen

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 7 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Document the fact that this method `getNeighboringOnGenomeCigarElement()` iterates until it finds an "on-genome" operator.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: Maybe rename to `getNearestOnGenomeCigarElement()` to make this clearer

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 16]
```diff
@@ -160,6 +162,83 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
...
+
+        //after the first 'op' from the list
+        state.stepForwardOnGenome();
+        final PileupElement pe2 = state.makePileupElement();
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.PREV), ops.get(ops.size()-1));
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.NEXT), null);
+        for (final CigarOperator op : CigarOperator.values()){
+            Assert.assertFalse(pe2.isImmediatelyBefore(op), "immediately before " + op);
+            Assert.assertEquals(pe2.isImmediatelyAfter(op), op ==  ops.get(ops.size()-1), "immediately after " + op);
+        }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: This test is really hard to follow and poorly designed -- can you completely replace it with a new test that:

-constructs PileupElements directly using a read that has a cigar with all possible cigar elements (don't use AlignmentStateMachine, please!), and various offsets into that read
-does direct, straightforward assertions on these PileupElements of the form: `pe.isImmediatelyBefore(CigarOperator.S)`, etc.
-includes test cases in which the current offset is not at the beginning/end of the current cigar element

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 17]
```diff
@@ -160,6 +162,83 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
...
+                    tests.add(new Object[]{read, middleOp});
+                }
+            }
+        }
+
+        return tests.toArray(new Object[][]{});
+    }
+
+    @Test(dataProvider = "PrevAndNextTest_simple")
+    public void testImmediateBeforeAndAfterTest_simple(final GATKRead read, final CigarOperator middleOp) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Same comment as above -- redesign to construct `PileupElements` directly rather than confusing things by going through `AlignmentStateMachine`

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java [hunk 18]
```diff
@@ -328,6 +329,13 @@ public void testSimplePileupWithIndelsOffset(){
         final int offset = 4;
         final ReadPileup pu = new ReadPileup(loc, reads, offset);
 
+        final PileupElement firstElem = pu.getElementForRead(read1);
+        Assert.assertNull(firstElem.getAdjacentOperator(PileupElement.Direction.NEXT));
+        Assert.assertNull(firstElem.getAdjacentOperator(PileupElement.Direction.PREV));
+        final PileupElement secondElem = pu.getElementForRead(read2);
+        Assert.assertEquals(secondElem.getAdjacentOperator(PileupElement.Direction.NEXT), CigarOperator.I);
+        Assert.assertNull(secondElem.getAdjacentOperator(PileupElement.Direction.PREV));
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/ReadPileupUnitTest.java]: Replace with high-quality direct testing on `getAdjacentOperator()` in `PileupElementUnitTest`

**Comment by droazen** [comment]
Back to @akiezun. Code itself looks good now, just need to rename `getNeighboringOnGenomeCigarElement()`. Tests are TERRIBLE though and need to be re-written from scratch.

**Comment by akiezun** [comment]
these things are super hard to create by hand though. I may just abandon this PR - the speedup may not be worth the effort here. I'll see

**Comment by droazen** [comment]
It's not super-hard -- just create a `DataProvider` with a single read containing all the possible cigar elements. Then each test case will be: a `PileupElement` constructed from that read, an offset, expected immediately before element, and expected immediately after element. Super easy!

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: renamed getNeighboringOnGenomeCigarElement  -> getNearestOnGenomeCigarElement

**Comment by akiezun** [comment]
added some more direct tests (sorry about the whitespace changes). back to @droazen

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java [hunk 19]
```diff
@@ -378,16 +398,16 @@ public CigarElement getPreviousOnGenomeCigarElement() {
...
      */
     public CigarElement getNextOnGenomeCigarElement() {
-        return getNeighboringOnGenomeCigarElement(Direction.NEXT);
+        return getNearestOnGenomeCigarElement(Direction.NEXT);
     }
 
     /**
      * Helper function to get the cigar element of the next or previous genomic position
      * @param direction the direction to look in
      * @return a CigarElement, or null if no such element exists
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: `a CigarElement` -> `nearest on-genome CigarElement`

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 20]
```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
...
+                Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+            }
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_afterFirstD() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 10);
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Also assert that `pe.atStartOfCurrentCigar()` is true

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 21]
```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
...
+            }
+        }
+
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op));
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_insideDs() {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: This test case is not "inside" the deletion, is it? Isn't it inside the second M operator instead, since the deletion consumes no bases in the bases array?

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 22]
```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
...
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+        }
+    }
+
+    @DataProvider(name = "elementData")
+    private Object[][] elementData() {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Use a more descriptive name for your `DataProvider` -- eg., `adjacentElementTestData` or something similar

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 23]
```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
...
+    public void testOperators(GATKRead read, int goodOffset, boolean atStartOfCurrentCigar, boolean atEndOfCurrentCigar, CigarOperator prev, CigarOperator next) throws Exception {
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, goodOffset);
+        Assert.assertEquals(pe.atStartOfCurrentCigar(), atStartOfCurrentCigar, "atStartOfCurrentCigar");
+        Assert.assertEquals(pe.atEndOfCurrentCigar(), atEndOfCurrentCigar, "atEndOfCurrentCigar");
+        Assert.assertEquals(pe.getAdjacentOperator(PileupElement.Direction.PREV), prev, "prev");
+        Assert.assertEquals(pe.getAdjacentOperator(PileupElement.Direction.NEXT), next, "next");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertEquals(pe.isImmediatelyAfter(op), atStartOfCurrentCigar && op == prev);
+            Assert.assertEquals(pe.isImmediatelyBefore(op), atEndOfCurrentCigar && op == next);
+        }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: No need for a loop here -- replace with two ifs for clarity:

    if ( atStartOfCurrentCigar ) {
        Assert.assertTrue(pe.isImmediatelyAfter(prev);
    }

    if ( atEndOfCurrentCigar ) {
        Assert.assertTrue(pe.isImmediatelyBefore(next);
    }

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 24]
```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
...
+                {read, 35, false, false, X, S},
+                {read, 36, false, false, X, S},
+                {read, 37, false, false, X, S},
+                {read, 38, false, false, X, S},
+                {read, 39, false, true, X, S},
+        };
+    }
+
+    @Test(dataProvider = "elementData")
+    public void testOperators(GATKRead read, int goodOffset, boolean atStartOfCurrentCigar, boolean atEndOfCurrentCigar, CigarOperator prev, CigarOperator next) throws Exception {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: `testAdjacentElements` might be a better name

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 23 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: i do want to check that it says `false` for other operators though

**Comment by droazen** [comment]
Back to @akiezun -- minor remaining comments, then squash/merge

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 23 above)
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Ah yes, good point.

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java [hunk 25]
```diff
@@ -160,17 +162,94 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
         Assert.assertEquals(pe2.getNextOnGenomeCigarElement(), null);
     }
 
+    @Test(dataProvider = "PrevAndNextTest")
+    public void testImmediateBeforeAndAfterTest(final GATKRead read, final CigarOperator firstOp, final CigarOperator lastOp, final List<CigarOperator> ops) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: Do you still want to keep these given the new tests? Do they have independent value?

File: src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java (hunk 19 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/pileup/PileupElement.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 20 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 21 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: correct

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 22 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 24 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java (hunk 25 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java]: i think they are fine to keep, very little cost to them
